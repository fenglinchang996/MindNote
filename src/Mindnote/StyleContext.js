import { createContext } from "react";
import { STROKE_TYPE } from "./utils/enums";

const styleContext = createContext({
  defaultNodeStyle: {
    width: 120,
    height: 40,
    rxRatio: 0.1,
    ryRatio: 0.2,
    borderType: STROKE_TYPE.SOLID,
    style: {
      fill: "#ffffff",
      stroke: "#000000",
      strokeWidth: 2,
      strokeLinecap: "round",
      strokeDasharray: "none",
    },
  },
  nodeStyles: [
    {
      rxRatio: 0.1,
      ryRatio: 0.2,
      borderType: STROKE_TYPE.SHORT_DASH,
      style: {
        fill: "#ffffff",
        stroke: "#008000",
        strokeWidth: 2,
        strokeLinecap: "round",
        strokeDasharray: [4, 4],
      },
    },
    {
      rxRatio: 0.1,
      ryRatio: 0.5,
      borderType: STROKE_TYPE.LONG_DASH,
      style: {
        fill: "#ffffff",
        stroke: "#0000ff",
        strokeWidth: 2,
        strokeLinecap: "round",
        strokeDasharray: [8, 4],
      },
    },
  ],
  nodePointStyle: {
    r: 5,
    style: {
      stroke: "#000000",
      fill: "#ffffff",
      cursor: "pointer",
    },
  },
  nodeContentStyle: {
    widthRatio: 0.9,
    heightRatio: 0.8,
    style: {
      fontSize: "0.8rem",
      width: "100%",
      height: "100%",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      backgroundColor: "transparent",
    },
  },
  virtualNodeStyle: {
    style: {
      stroke: "#ffc0cb",
      strokeDasharray: [5, 2],
    },
  },
  selectedNodeStyle: {
    style: {
      stroke: "#ff0000",
    },
  },
  viewNodeStyle: {
    style: {
      stroke: "#ff0000",
    },
  },
  connectionArrowStyle: {
    style: {},
  },
  defaultCurveStyle: {
    type: STROKE_TYPE.SOLID,
    style: {
      fill: "none",
      stroke: "#0000ff",
      strokeWidth: 5,
      strokeLinecap: "round",
      strokeDasharray: "none",
    },
  },
  curveStyles: [
    null,
    {
      type: STROKE_TYPE.SHORT_DASH,
      style: {
        fill: "none",
        stroke: "#ffbf00",
        strokeWidth: 3,
        strokeLinecap: "round",
        strokeDasharray: [6, 6],
      },
    },
    {
      type: STROKE_TYPE.SHORT_DASH,
      style: {
        fill: "none",
        stroke: "#800080",
        strokeWidth: 5,
        strokeLinecap: "round",
        strokeDasharray: [20, 10],
      },
    },
  ],
  defaultCurveArrowStyle: {
    style: {
      fill: "#0000ff",
    },
  },
  curveArrowStyles: [
    {
      style: {
        fill: "#0000ff",
      },
    },
  ],
  virtualCurveStyle: {
    style: {},
  },
  selectedCurveStyle: {
    style: {},
  },
  viewCurveStyle: {
    style: {},
  },
  curvePointStyle: {
    r: 5,
    style: {
      stroke: "#000000",
      fill: "#ffffff",
      cursor: "pointer",
    },
  },
  curveControlStyle: {
    r: 5,
    style: {
      stroke: "#000000",
      fill: "#ffffff",
      cursor: "pointer",
    },
  },
  curveControlLineStyle: {
    style: {
      fill: "none",
      stroke: "#d3d3d3",
      strokeWidth: 5,
      strokeDasharray: "2, 2",
    },
  },
  noteStyle: {
    defaultWidth: 300,
    minWidth: 250,
  },
});

export default styleContext;
