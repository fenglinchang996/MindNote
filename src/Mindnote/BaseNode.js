import React, { useState, useEffect, useContext } from "react";
import StyleContext from "./StyleContext";
import ItemContext from "./ItemContext";
import SVGContext from "./SVGContext";
import { ITEM_TYPE } from "./enums";
import { EDGE } from "./enums";

const BaseNode = (props) => {
  const { nodeData, isFocused } = props;
  const {
    id,
    center,
    width,
    height,
    style,
    corners,
    edgeCenters,
    connections,
    parentId,
    upstreamCurveId,
    childrenId,
    downstreamCurvesId,
    top,
    right,
    left,
    bottom,
  } = nodeData;
  const { connectionArrowStyle } = useContext(StyleContext);
  const { drawNewNode } = useContext(SVGContext);

  // Origin
  const [origin, setOrigin] = useState({ x: 0, y: 0 });
  // Calculate and set Node origin
  useEffect(() => {
    setOrigin({ x: center.x - 0.5 * width, y: center.y - 0.5 * height });
  }, [center, width, height]);

  // Connection Arrow
  const connectionBlockLength = 20;
  const connectionArrows = {
    top: {
      x: connections.topConnection.x - 0.5 * connectionBlockLength,
      y: connections.topConnection.y - connectionBlockLength,
      width: connectionBlockLength,
      height: connectionBlockLength,
    },
    right: {
      x: connections.rightConnection.x,
      y: connections.rightConnection.y - 0.5 * connectionBlockLength,
      width: connectionBlockLength,
      height: connectionBlockLength,
    },
    bottom: {
      x: connections.bottomConnection.x - 0.5 * connectionBlockLength,
      y: connections.bottomConnection.y,
      width: connectionBlockLength,
      height: connectionBlockLength,
    },
    left: {
      x: connections.leftConnection.x - connectionBlockLength,
      y: connections.leftConnection.y - 0.5 * connectionBlockLength,
      width: connectionBlockLength,
      height: connectionBlockLength,
    },
  };

  return (
    <g>
      <rect
        x={origin.x}
        y={origin.y}
        width={width}
        height={height}
        style={style}
      ></rect>
      <g display={isFocused ? "block" : "none"}>
        {connectionArrows ? (
          <>
            <ConnectionArrow
              arrowData={connectionArrows.top}
              fa="caret-up"
              drawNewNode={(e) => drawNewNode(e, id, EDGE.TOP)}
            />
            <ConnectionArrow
              arrowData={connectionArrows.right}
              fa="caret-right"
              drawNewNode={(e) => drawNewNode(e, id, EDGE.RIGHT)}
            />
            <ConnectionArrow
              arrowData={connectionArrows.bottom}
              fa="caret-down"
              drawNewNode={(e) => drawNewNode(e, id, EDGE.BOTTOM)}
            />
            <ConnectionArrow
              arrowData={connectionArrows.left}
              fa="caret-left"
              drawNewNode={(e) => drawNewNode(e, id, EDGE.LEFT)}
            />
          </>
        ) : (
          ""
        )}
      </g>
    </g>
  );
};

const ConnectionArrow = (props) => {
  const { arrowData, fa, drawNewNode } = props;
  return (
    <foreignObject {...arrowData} cursor="pointer" onMouseDown={drawNewNode}>
      <div className="connection-arrow">
        <i className={`fas fa-${fa}`}></i>
      </div>
    </foreignObject>
  );
};

export default BaseNode;
