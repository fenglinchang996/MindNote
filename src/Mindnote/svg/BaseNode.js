import React, { useContext } from "react";
import StyleContext from "../StyleContext";

const BaseNode = (props) => {
  const { nodeData } = props;
  const { title, center, width, height, style } = nodeData;
  const { nodeContentStyle } = useContext(StyleContext);

  // Origin
  const origin = { x: center.x - 0.5 * width, y: center.y - 0.5 * height };

  return (
    <g>
      <rect
        x={origin.x}
        y={origin.y}
        width={width}
        height={height}
        style={style}
      ></rect>
      <foreignObject
        x={origin.x + ((1 - nodeContentStyle.widthRatio) / 2) * width}
        y={origin.y + ((1 - nodeContentStyle.heightRatio) / 2) * height}
        width={nodeContentStyle.widthRatio * width}
        height={nodeContentStyle.heightRatio * height}
      >
        <div style={nodeContentStyle.style}>
          <div style={{ textAlign: "center" }}>
            {title ? title : "Click to Add Note"}
          </div>
        </div>
      </foreignObject>
    </g>
  );
};
export default BaseNode;
