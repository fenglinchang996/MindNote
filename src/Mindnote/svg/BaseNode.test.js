import React from "react";
import renderer from "react-test-renderer";
import StyleContext from "../StyleContext";
import BaseNode from "./BaseNode";

test("BaseNode", () => {
  const centerNodeData = {
    width: 120,
    height: 40,
    center: { x: 480, y: 270 },
    title: "Center",
    level: 0,
  };
  const StyleContextValue = {
    nodeStyles: [],
    defaultNodeStyle: {
      rxRatio: 0.1,
      ryRatio: 0.2,
      style: {
        fill: "white",
        stroke: "black",
        strokeWidth: 2,
      },
    },
    nodeContentStyle: {
      widthRatio: 0.9,
      heightRatio: 0.8,
      style: {
        fontSize: "0.8rem",
        width: "100%",
        height: "100%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        backgroundColor: "transparent",
      },
    },
  };
  const component = renderer.create(
    <StyleContext.Provider value={StyleContextValue}>
      <BaseNode nodeData={centerNodeData} />
    </StyleContext.Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
