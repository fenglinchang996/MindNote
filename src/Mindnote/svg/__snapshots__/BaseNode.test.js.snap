// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseNode 1`] = `
<g>
  <rect
    height={40}
    rx={12}
    ry={8}
    width={120}
    x={420}
    y={250}
  />
  <foreignObject
    height={32}
    width={108}
    x={426}
    y={254}
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "display": "flex",
          "fontSize": "0.8rem",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Center
      </div>
    </div>
  </foreignObject>
</g>
`;
